### PR TITLE
Feature/sort enforcement refactor

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Which issue does this PR close?

Closes #.

# Rationale for this change

As suggested by @mingmwang in the [discussion](https://github.com/apache/arrow-datafusion/pull/5171#discussion_r1099988290). Comparing ordering between schemas when there are multiple executors in between them may produce incorrect results. Although this situation may not arise in practice with the current ordering of the optimization rules. We have changed the implementation in case one encounters it (Nobody knows).

# What changes are included in this PR?

With this change parallelize sort rule compares the immediate table for converting `Coalesce + Executors that do not require SingleDistribution + Sort` cascades to `Executors that do not require SingleDistribution + Sort + SortPreservingMerge` cascade.
We also added a check that the final Sort in the cascade rule is actually a Global sort. 
# Are these changes tested?

We added a unit test to verify that rule only ends when Sort is a global sort. Also, the existing [test](https://github.com/synnada-ai/arrow-datafusion/blob/157e204bb35e5e1466b8d6c11846e1a15ef048e5/datafusion/core/src/physical_optimizer/sort_enforcement.rs#L1680) checks the sort comparison is done between immediate executors.

# Are there any user-facing changes?

### CODE REVIEW

**Comment by ozankabak** [comment]
This is the first follow-on PR to address the minor issues that were discussed in the original PR (#5171).

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs [hunk 1]
```diff
@@ -1776,6 +1745,47 @@ mod tests {
...
+            "        CoalescePartitionsExec",
+            "          RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=0",
+            "            MemoryExec: partitions=0, partition_sizes=[]",
+        ];
+        let expected_optimized = vec![
+            "SortPreservingMergeExec: [nullable_col@0 ASC]",
+            "  SortExec: [nullable_col@0 ASC]",
+            "    RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=10",
+            "      RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=0",
+            "        MemoryExec: partitions=0, partition_sizes=[]",
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/sort_enforcement.rs:1783]: This plan definitely looks better than the input.

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs [hunk 2]
```diff
@@ -1718,8 +1687,8 @@ mod tests {
         ];
         let expected_optimized = vec![
             "SortPreservingMergeExec: [nullable_col@0 ASC]",
-            "  FilterExec: NOT non_nullable_col@1",
-            "    SortExec: [nullable_col@0 ASC]",
+            "  SortExec: [nullable_col@0 ASC]",
+            "    FilterExec: NOT non_nullable_col@1",
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/sort_enforcement.rs:1691]: This plan change looks better to me as well (do filtering before sort)

**Review by alamb — APPROVED** [review]
I am not an expert in this code, so I would really like @mingmwang to review as well prior to merge.

However, I went over the plan changes and new test and they all make sense to me and the fact that the existing tests all pass means I am 👍  for this PR.

Thank you @mustafasrepo 

cc @crepererum  who may also be interested in this PR as we are reworking some of our plan construction in IOx as well

**Comment by mingmwang** [comment]
I will review this PR carefully today.

**Comment by mingmwang** [comment]
The change LGTM.

**Pull request merged**